### ISSUE TITLE
[Feature Request]: add EXEC to Dockerfile syntax

### ISSUE THREAD

**Issue by inlineHamed** [moby/moby#29706]
There are cases in dockerfile that we want to do something while the entrypoint is running so it has the the effect of exec after running container. some example of these cases are:
1) in MySQL doing any mysql command like executing a .sql file
2) in Solr doing config with solr command

currently the approach is to use something like this: (as suggested in http://stackoverflow.com/questions/25920029/setting-up-mysql-and-importing-dump-within-dockerfile)
```
RUN /bin/bash -c "/usr/bin/mysqld_safe &" && \
    sleep 5 && \
    mysql -u root -e "CREATE DATABASE mydb"
```

but it has some disadvantages:
1) finding entrypoint and cmd is not straightforward.
2) entrypoint or cmd might change.
3) sleep part is not safe as it might needs more time based on host performance.
3) it's not a clean code.

My suggestion is to add EXEC command to Dockerfile so above smaple will be like this:

`EXEC sleep 5 && mysql -u root -e "CREATE DATABASE mydb"`

or maybe we could remove "sleep 5" by using healthcheck feature:

`EXEC mysql -u root -e "CREATE DATABASE mydb"`

**Comment by hmdhasani** [comment]
no one on this? :(

**Comment by thaJeztah** [comment]
If I understand correctly, your goal is to create a MySQL image, with a "default" database inside the image, so that users can start the image, and have the default database?

> sleep part is not safe as it might needs more time based on host performance.

Correct, sleep is not the correct solution; you should have a script that checks if the database is available to take requests, and retry otherwise.

> finding entrypoint and cmd is not straightforward.
> My suggestion is to add EXEC command to Dockerfile so above smaple will be like this:

I don't think this will solve much; again, if I understand correctly, the only reason for adding this would be so that there's no need to check what the `entrypoint` of the image is? But if the entrypoint of the image _changes_, chances are big that your `EXEC` also needs to be updated. "Hiding" this will only result in possible bugs in your Dockerfile.

Note that the official MySQL image already allows you to ship with a SQL-dump to initialise the database on first run; if you add a `.sql` or `.sh` file to `/docker-entrypoint-initdb.d` inside your custom image, then the image will load that file the first time the image is run; see the [entrypoint script](https://github.com/docker-library/mysql/blob/4dd33136c4739667a223d39b6f829beb27b235cf/8.0/docker-entrypoint.sh#L76-L173) and the README of that image; https://hub.docker.com/_/mysql/

In other words; a Dockerfile like this would do what you want;

```
FROM mysql
COPY ./my-database-dump.sql /docker-entrypoint-initdb.d/
```

**Comment by hmdhasani** [comment]
@thaJeztah this is only one of my needs and there are many other cases that a server like app must be running to be configured or initialized.
There are 2 downside for that kind of solutions:
1) We must relay on publisher solution for that kind of needs. currently most of server likes images (postgres, solr, redis, cassandra, sentr ,... ) does not provide such a solution and MySQL did it recently.
2) The docker-entrypoint-initdb.d does not fit efficiently as it put the process in container run time so for example if my .sql file take 1 minutes to run, it will take this 1 minute every time that a new container run but doing it on build images time will reduce it only on image build and also take advantages of caching.

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
